static/widgets/compiler-picker-popup.ts

```diff
@@ -0,0 +1,266 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import $ from 'jquery';
+
+import * as sifter from '@orchidjs/sifter';
+
+import {CompilerInfo} from '../../types/compiler.interfaces';
+import {intersection, remove, unique} from '../../lib/common-utils';
+import {unwrap, unwrapString} from '../assert';
+import {CompilerPicker} from './compiler-picker';
+import {CompilerService} from '../compiler-service';
+import {highlight} from '../highlight';
+
+export class CompilerPickerPopup {
+    modal: JQuery<HTMLElement>;
+    searchBar: JQuery<HTMLElement>;
+    architectures: JQuery<HTMLElement>;
+    compilerTypes: JQuery<HTMLElement>;
+    compilersContainer: JQuery<HTMLElement>;
+    resultsContainer: JQuery<HTMLElement>;
+    favoritesContainer: JQuery<HTMLElement>;
+
+    groups: {value: string; label: string}[];
+    options: (CompilerInfo & {$groups: string[]})[];
+    langId: string;
+
+    isaFilters: string[];
+    categoryFilters: string[];
+
+    sifter: sifter.Sifter;
+    searchResults: ReturnType<sifter.Sifter['search']> | undefined;
+
+    constructor(private readonly compilerPicker: CompilerPicker) {
+        this.modal = $('#compiler-picker-modal').clone(true);
+        this.searchBar = this.modal.find('.compiler-search');
+        this.architectures = this.modal.find('.architectures');
+        this.compilerTypes = this.modal.find('.compiler-types');
+        this.compilersContainer = this.modal.find('.compilers-row');
+        this.resultsContainer = this.modal.find('.compilers');
+        this.favoritesContainer = this.modal.find('.favorites');
+
+        this.modal.on('shown.bs.modal', () => {
+            this.searchBar[0].focus();
+        });
+    }
+
+    setLang(groups: {value: string; label: string}[], options: (CompilerInfo & {$groups: string[]})[], langId: string) {
+        this.groups = groups;
+        this.options = options;
+        this.langId = langId;
+        this.setupFilters();
+        this.sifter = new sifter.Sifter(options, {
+            diacritics: false,
+        });
+    }
+
+    setupFilters() {
+        // get available instruction sets
+        const compilers = Object.values(this.compilerPicker.compilerService.getCompilersForLang(this.langId) ?? {});
+        // If instructionSet is '', just label it unknown
+        const instruction_sets = compilers.map(compiler => compiler.instructionSet || 'other');
+        this.architectures.empty();
+        this.architectures.append(
+            ...unique(instruction_sets)
+                .sort()
+                .map(isa => `<span class="architecture" data-value=${isa}>${isa}</span>`),
```

Comment:
We're probably ok but does `isa` etc all need to be escaped here and in the compilerTypes?

Could we use jquery's magic instead somehow (or the raw equivalent).
```js
isa => $("<span/>").addClass("architecture").data("value", isa).text(isa)
```

Reply:
thank you, very good catch!